internal/provider/resource_gitlab_personal_access_tokens.go

```diff
@@ -0,0 +1,242 @@
+				Required:    true,
+				ForceNew:    true,
+			},
+			"scopes": {
+				Description: fmt.Sprintf("The scope for the personal access token. It determines the actions which can be performed when authenticating with this token. Valid values are: %s.", renderValueListForDocs(validPersonalAccessTokenScopes)),
+				Type:        schema.TypeSet,
+				Required:    true,
+				ForceNew:    true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice(validPersonalAccessTokenScopes, false),
+				},
+			},
+			"active": {
+				Description: "True if the token is active.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"revoked": {
+				Description: "True if the token is revoked.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"created_at": {
+				Description: "Time the token has been created, RFC3339 format.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"expires_at": {
+				Description:      "The token expires at midnight UTC on that date. The date must be in the format YYYY-MM-DD. Default is never.",
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: isISO6801Date,
+			},
+			"token": {
+				Description: "The personal access token. This is only populated when creating a new personal access token. This attribute is not available for imported resources.",
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+			},
+		},
+	}
+})
+
+func resourceGitlabPersonalAccessTokenCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	userId := d.Get("user_id").(int)
+
+	currentUserAdmin, err := isCurrentUserAdmin(client)
+	if err != nil {
+		return diag.Errorf("[ERROR] cannot query the user API for current user: %v", err)
+	}
+
+	if !currentUserAdmin {
+		return diag.Errorf("[ERROR] Current user needs to be admin when creating a personal access token: %v", err)
+	}
+
+	options := &gitlab.CreatePersonalAccessTokenOptions{
+		Name:   gitlab.String(d.Get("name").(string)),
+		Scopes: stringSetToStringSlice(d.Get("scopes").(*schema.Set)),
+	}
+
+	log.Printf("[DEBUG] create gitlab PersonalAccessToken %s (scopes: %s) for user ID %d", *options.Name, options.Scopes, userId)
+
+	if v, ok := d.GetOk("expires_at"); ok {
+		parsedExpiresAt, err := time.Parse("2006-01-02", v.(string))
+		if err != nil {
+			return diag.Errorf("Invalid expires_at date: %v", err)
+		}
+
+		parsedExpiresAtISOTime := gitlab.ISOTime(parsedExpiresAt)
+		options.ExpiresAt = &parsedExpiresAtISOTime
+	}
+
+	personalAccessToken, _, err := client.Users.CreatePersonalAccessToken(userId, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tokenId := strconv.Itoa(personalAccessToken.ID)
+	userIdString := strconv.Itoa(userId)
+
+	d.SetId(buildTwoPartID(&userIdString, &tokenId))
```

Comment:
tbh. I'd favor:

```suggestion
	d.SetId(fmt.Sprintf("%d:%d", userID, personalAccessToken.ID))
```

Reply:
or refactor into own function like here:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/4337c7437f809c2190d3e07e94137fa58b734729/internal/provider/resource_gitlab_project_issue.go#L254

Reply:
Do we want to remove the helper function from `util.go` then and change all instances to the Sptrintf option?

Reply:
I'm not 100% sure yet, the function usually gives some more meaning to it and also consistently uses `:` as the delimiter. I just don't like the current verbosity ...

I suggest to use a separate function and there you may or may not use `buildTwoPartID`, but at least the resource CRUD function is concise. WDYT?

Reply:
Mmmm agreed on the verbosity. I was looking at golang and couldn't find anything like overloading functions or anonymous type inputs.  I'm for more concise with the Sprintf and then during reviews we can catch people doing non-standard ID's and such.  We could also potentially do a custom linting rule for this as well

Reply:
the current version is fine for me for now :)